scripts/scraper-ng/rules/html-require-recipe-ingredients.js

```diff
@@ -0,0 +1,212 @@
+      message.ruleId = `html-must-match-recipe:${recipeName}`;
+      message.fatal = true;
+    }
+  };
+}
+
+/**
+ * Functions to check for recipe ingredients in Kuma page sources.
+ *
+ * The key is the name of a recipe ingredient (e.g.,
+ * `data.browser_compatbiility` or `prose.syntax`) and the value is a function
+ * to process a tree and file.
+ *
+ * The functions must take two arguments: a hast tree and a VFile. The function
+ * must return `true` if the page contains the recipe ingredient, or `false` if
+ * not. The function may also log messages against the file.
+ *
+ */
+const ingredientHandlers = {
+  "data.browser_compatibility": (tree, file) => {
+    const id = "Browser_compatibility";
+    const ingredient = "data.browser_compatibility";
+    const body = select(`body`, tree);
+
+    const heading = select(`h2#${id}`, tree);
+    if (heading === null) {
+      warnMissingIngredient(file, ingredient);
+      return false;
+    }
+
+    const nextHeading = select(`#${id} ~ h2`, tree);
+    const sectionStart = body.children.indexOf(heading);
+    const sectionEnd =
+      nextHeading !== null
+        ? body.children.indexOf(nextHeading)
+        : body.children.length;
+    const subTree = {
+      type: "root",
+      children: body.children.slice(sectionStart, sectionEnd)
+    };
+
+    let macroCount = 0;
+    visit(
+      subTree,
+      node => isMacro(node, "compat"),
+      () => {
+        macroCount += 1;
+      }
+    );
+
+    if (macroCount !== 1) {
+      warnMissingIngredient(file, ingredient);
+      return false;
+    }
+    return true;
+  },
+  "data.examples": requireTopLevelHeading("data.examples", "Examples"),
+  "data.specifications": (tree, file) => {
+    const id = "Specifications";
+    const ingredient = "data.specifications";
+    const body = select(`body`, tree);
+
+    const heading = select(`h2#${id}`, tree);
+    if (heading === null) {
+      warnMissingIngredient(file, ingredient);
+      return false;
+    }
+
+    const sectionStart = body.children.indexOf(heading);
+    const sectionEnd = body.children.indexOf(select(`#${id} ~ h2`, tree));
+    const subTree = {
+      type: "root",
+      children: body.children.slice(sectionStart, sectionEnd)
+    };
+
+    let sectionOk = false;
+    visit(subTree, "text", node => {
+      if (isMacro(node) && node.data.macroName === "specname") {
+        sectionOk = true;
+        return visit.SKIP;
+      }
+
+      if (node.value.includes("Not part of any standard")) {
+        sectionOk = true;
+        return visit.SKIP;
+      }
+    });
+
+    if (!sectionOk) {
+      warnMissingIngredient(file, ingredient);
+      return false;
+    }
+    return true;
+  },
+  "prose.description": requireTopLevelHeading(
+    "prose.description",
+    "Description"
+  ),
+  "prose.see_also": requireTopLevelHeading("prose.see_also", "See_also"),
+  "prose.short_description": (tree, file) => {
+    if (select("body > p", tree) === null) {
```

Comment:
This is too liberal: a page like https://wiki.developer.mozilla.org/en-US/docs/User:wbamberg/test-linter2 passes. I think testing this is tricky: you more or less want the first element of the raw document body to be `<p>`, except that sidebars will usually precede this, and pages very occasionally (and wrongly) include the interactive example macro in `<p>` instead of `<div>` (luckily this is very apparent to humans because it means the example wraps at text-width, instead of getting the full page width).

Reply:
Pushing something that hopefully does a little better at this